core/servers/chat/mrc_multiplexer.js

```diff
@@ -0,0 +1,269 @@
+            // this.log.debug('Respond to heartbeat');
+            this.sendToMrcServer(socket, 'CLIENT', '', 'SERVER', 'ALL', '', `IMALIVE:${siteName}`);
+
+        } else {
+            // if not a heartbeat, and we have clients then we need to send something to them
+            this.sendToClient(message);
+        }
+    }
+
+    /**
+     * Takes an MRC message and parses it into something usable
+     */
+    parseMessage(line) {
+        const msg = line.split('~');
+        if (msg.length < 7) {
+            return;
+        }
+
+        return {
+            from_user: msg[0],
+            from_site: msg[1],
+            from_room: msg[2],
+            to_user: msg[3],
+            to_site: msg[4],
+            to_room: msg[5],
+            body: msg[6]
+        };
+    }
+
+    /**
+     * Receives a message from a local client and sanity checks before sending on to the central MRC server
+     */
+    receiveFromClient(username, message) {
+        try {
+            message = JSON.parse(message);
+        } catch (e) {
+            Log.debug({ server : 'MRC', user : username, message : message }, 'Dodgy message received from client');
+        }
+
+        this.sendToMrcServer(mrcCentralConnection, message.from_user, message.from_room, message.to_user, message.to_site, message.to_room, message.body);
+    }
+
+    /**
+     * Converts a message back into the MRC format and sends it to the central MRC server
+     */
+    sendToMrcServer(socket, fromUser, fromRoom, toUser, toSite, toRoom, messageBody) {
+        const config = Config();
+        const siteName = slugify(config.general.boardName);
+
+        const line = [
+            fromUser,
+            siteName,
+            sanitiseRoomName(fromRoom),
+            sanitiseName(toUser || ''),
+            sanitiseName(toSite || ''),
+            sanitiseRoomName(toRoom || ''),
+            sanitiseMessage(messageBody)
+        ].join('~') + '~';
+
+        Log.debug({ server : 'MRC', data : line }, 'Sending data');
+        return socket.write(line + '\n');
+    }
+};
+
+/**
+ * User / site name must be ASCII 33-125, no MCI, 30 chars max, underscores
+ */
+function sanitiseName(str) {
+    return str.replace(
+        /\s/g, '_'
+    ).replace(
+        /[^\x21-\x7D]|(\|\w\w)/g, '' // Non-printable & MCI
+    ).substr(
+        0, 30
+    );
+}
+
+function sanitiseRoomName(message) {
+    return message.replace(/[^\x21-\x7D]|(\|\w\w)/g, '').substr(0, 30);
+}
+
+function sanitiseMessage(message) {
+    return message.replace(/[^\x20-\x7D]/g, '');
+}
+
+/**
+ * SLugifies the BBS name for use as an MRC "site name"
+ */
+function slugify(text) {
+    return text.toString()
+        .replace(/\s+/g, '_')           // Replace spaces with _
+        .replace(/[^\w\-]+/g, '')       // Remove all non-word chars
+        .replace(/\-\-+/g, '_')         // Replace multiple - with single -
+        .replace(/^-+/, '')             // Trim - from start of text
```

Comment:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim

Reply:
I'm missing how `trim` is useful here?

Reply:
Nope sorry, I needed to look at the regex closer - you're removing '-', not spaces.